db/blob/blob_source.cc

```diff
@@ -152,6 +153,142 @@ Status BlobSource::GetBlob(const ReadOptions& read_options,
   return s;
 }
 
+void BlobSource::MultiGetBlob(
+    const ReadOptions& read_options,
+    const autovector<std::reference_wrapper<const Slice>>& user_keys,
+    uint64_t file_number, uint64_t file_size,
+    const autovector<uint64_t>& offsets,
+    const autovector<uint64_t>& value_sizes, autovector<Status*>& statuses,
+    autovector<PinnableSlice*>& blobs, uint64_t* bytes_read) {
+  size_t num_blobs = user_keys.size();
+  assert(num_blobs > 0);
+  assert(num_blobs <= MultiGetContext::MAX_BATCH_SIZE);
+  assert(num_blobs == offsets.size());
+  assert(num_blobs == value_sizes.size());
+  assert(num_blobs == statuses.size());
+  assert(num_blobs == blobs.size());
+
+#ifndef NDEBUG
+  for (size_t i = 0; i < offsets.size() - 1; ++i) {
+    assert(offsets[i] <= offsets[i + 1]);
+  }
+#endif  // !NDEBUG
+
+  using Mask = uint64_t;
+  Mask cache_hit_mask_ = 0;
+
+  Status s;
+  uint64_t total_bytes = 0;
+  const OffsetableCacheKey base_cache_key(db_id_, db_session_id_, file_number,
+                                          file_size);
+
+  if (blob_cache_) {
+    size_t cached_blob_count = 0;
+    for (size_t i = 0; i < num_blobs; ++i) {
+      CachableEntry<std::string> blob_entry;
+      const CacheKey cache_key = base_cache_key.WithOffset(offsets[i]);
+      const Slice key = cache_key.AsSlice();
+
+      s = GetBlobFromCache(key, &blob_entry);
+      if (!s.ok()) {
+        assert(statuses[i]);
+        *statuses[i] = s;
+        continue;
+      }
+
+      if (blob_entry.GetValue()) {
+        assert(blob_entry.GetValue()->size() == value_sizes[i]);
+        blobs[i]->PinSelf(*blob_entry.GetValue());
+
+        // Update the counter for the number of valid blobs read from the cache.
+        ++cached_blob_count;
+        total_bytes += value_sizes[i];
+        cache_hit_mask_ |= (Mask{1} << i);  // cache hit
+      }
+    }
+
+    // All blobs were read from the cache.
+    if (cached_blob_count == num_blobs) {
+      if (bytes_read) {
+        *bytes_read = total_bytes;
+      }
+      return;
+    }
+  }
+
+  const bool no_io = read_options.read_tier == kBlockCacheTier;
+  if (no_io) {
+    for (size_t i = 0; i < num_blobs; ++i) {
+      if (!(cache_hit_mask_ & (Mask{1} << i))) {
+        assert(statuses[i]);
+        *statuses[i] =
+            Status::Incomplete("Cannot read blob(s): no disk I/O allowed");
+      }
+    }
+    return;
+  }
+
+  {
+    // Find the rest of blobs from the file since I/O is allowed.
+    autovector<std::reference_wrapper<const Slice>> _user_keys;
+    autovector<uint64_t> _offsets;
+    autovector<uint64_t> _value_sizes;
+    autovector<Status*> _statuses;
+    autovector<PinnableSlice*> _blobs;
+    uint64_t _bytes_read = 0;
+
+    for (size_t i = 0; i < num_blobs; ++i) {
+      if (!(cache_hit_mask_ & (Mask{1} << i))) {
```

Comment:
It seems that it's possible that `!statuses[i].ok() && !(cache_hit_mask_ & (Mask{1} << i))`. Therefore, line 255 can overwrite a previous non-ok status.

Reply:
Would be nice to cover this case in test too.

Reply:
Bring this up in case the `outdated` tag confuses us. The question is still valid.

Reply:
done. thanks! @riversand963

Reply:
SGTM